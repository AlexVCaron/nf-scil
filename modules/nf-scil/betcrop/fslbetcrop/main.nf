
process BETCROP_FSLBETCROP {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.6.0.sif':
        'scilus/scilus:1.6.0' }"

    input:
        tuple val(meta), path(image), path(bval), path(bvec), path(template), path(map)

    output:
        tuple val(meta), path("*_bet_cropped.nii.gz")            , emit: image
        tuple val(meta), path("*_bet_cropped_mask.nii.gz")       , emit: mask
        tuple val(meta), path("*_boundingBox.pkl")               , emit: bbox
        path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def b0_thr = task.ext.b0_thr ? "--b0_thr " + task.ext.b0_thr : ""
    def bet_dwi_f = task.ext.bet_dwi_f ? "-f " + task.ext.bet_dwi_f : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    if [[ -v $bval && -v $bvec ]]
    then
        scil_extract_b0.py $image $bval $bvec ${prefix}__b0.nii.gz --mean \
            $b0_thr --force_b0_threshold
        bet ${prefix}__b0.nii.gz ${prefix}__b0_bet.nii.gz -m -R $bet_dwi_f
        scil_image_math.py convert ${prefix}__b0_bet_mask.nii.gz ${prefix}__b0_bet_mask.nii.gz --data_type uint8 -f
        mrcalc $image ${prefix}__b0_bet_mask.nii.gz -mult ${prefix}__dwi_bet.nii.gz -quiet -nthreads 1

        scil_crop_volume.py $image ${prefix}__dwi_bet_cropped.nii.gz -f \
            --output_bbox ${prefix}__dwi_boundingBox.pkl -f
        scil_crop_volume.py ${prefix}__b0_bet_mask.nii.gz ${prefix}__dwi_bet_cropped_mask.nii.gz -f\
            --input_bbox ${prefix}__dwi_boundingBox.pkl -f
        scil_image_math.py convert ${prefix}__dwi_bet_cropped_mask.nii.gz ${prefix}__dwi_bet_cropped_mask.nii.gz \
            --data_type uint8 -f
    
    else
        export ANTS_RANDOM_SEED=1234

        antsBrainExtraction.sh -d 3 -a $image -e $template\
            -o bet/ -m $map -u 0
        scil_image_math.py convert bet/BrainExtractionMask.nii.gz ${prefix}__t1_bet_mask.nii.gz --data_type uint8
        mrcalc $image ${prefix}__t1_bet_mask.nii.gz -mult ${prefix}__t1_bet.nii.gz -nthreads 1

        scil_crop_volume.py $image ${prefix}__t1_bet_cropped.nii.gz\
            --output_bbox t1_boundingBox.pkl -f
        scil_crop_volume.py ${prefix}__t1_bet_mask.nii.gz ${prefix}__t1_bet_mask_cropped.nii.gz\
            --input_bbox t1_boundingBox.pkl -f
        scil_image_math.py convert ${prefix}__t1_bet_mask_cropped.nii.gz ${prefix}__t1_bet_mask_cropped.nii.gz --data_type uint8 -f

    fi     

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
        mrtrix: \$(mrcalc -version 2>&1 | sed -n 's/== mrcalc \\([0-9.]\\+\\).*/\\1/p')
        fsl: \$(flirt -version 2>&1 | sed -n 's/FLIRT version \\([0-9.]\\+\\)/\\1/p')
        ants: 2.4.3

    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_extract_b0.py -h
    bet -h
    scil_image_math.py -h
    mrcalc -h
    scil_crop_volume.py -h

    touch ${prefix}__dwi_bet_cropped.nii.gz
    touch ${prefix}__dwi_bet_cropped_mask.nii.gz
    touch ${prefix}__dwi_boundingBox.pkl

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
        mrtrix: \$(mrcalc -version 2>&1 | sed -n 's/== mrcalc \\([0-9.]\\+\\).*/\\1/p')
        fsl: \$(flirt -version 2>&1 | sed -n 's/FLIRT version \\([0-9.]\\+\\)/\\1/p')
        ants: 2.4.3
    END_VERSIONS
    """
}
