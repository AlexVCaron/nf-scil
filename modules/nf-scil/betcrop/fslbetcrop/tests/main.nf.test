nextflow_process {

    name "Test Process BETCROP_FSLBETCROP"
    script "../main.nf"
    process "BETCROP_FSLBETCROP"

    tag "modules"
    tag "modules_nfcore"
    tag "betcrop"
    tag "betcrop/fslbetcrop"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "others.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("betcrop - fslbetcrop - dwi") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.test_data['betcrop']['fslbetcrop']['dwi'], checkIfExists: true),
                    file(params.test_data['betcrop']['fslbetcrop']['bval'], checkIfExists: true),
                    file(params.test_data['betcrop']['fslbetcrop']['bvec'], checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("betcrop - fslbetcrop - t1") {

        config "./nextflow_t1.config"

        when {
            process {
                """                
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/t1.nii.gz"),
                        [],
                        []
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("betcrop - fslbetcrop - t1 - dilate") {

        config "./nextflow_t1_dilate.config"

        when {
            process {
                """                
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/t1.nii.gz"),
                        [],
                        []
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
