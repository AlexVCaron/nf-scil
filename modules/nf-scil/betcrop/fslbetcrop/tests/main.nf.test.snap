{
    "betcrop - fslbetcrop - dwi": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__b0_bet.nii.gz:md5,88d9cf75a5e270b893f20cac06aa194d",
                            "test__image_bet.nii.gz:md5,cf4cd471df9fa1d4e8b32fa13c2da016"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__b0_bet_mask.nii.gz:md5,23864b4bb35840287bb73c72529779a4",
                            "test__image_bet_mask.nii.gz:md5,bfe292b871bc66c2f7a6d38e71475af3"
                        ]
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,5336fac953810299feeb1a62e824ad35"
                ],
                "bbox": [
                    
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__b0_bet.nii.gz:md5,88d9cf75a5e270b893f20cac06aa194d",
                            "test__image_bet.nii.gz:md5,cf4cd471df9fa1d4e8b32fa13c2da016"
                        ]
                    ]
                ],
                "mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__b0_bet_mask.nii.gz:md5,23864b4bb35840287bb73c72529779a4",
                            "test__image_bet_mask.nii.gz:md5,bfe292b871bc66c2f7a6d38e71475af3"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5336fac953810299feeb1a62e824ad35"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-23T16:42:42.142214"
    },
    "betcrop - fslbetcrop - t1": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet.nii.gz:md5,687ccd83b5f7d49608e3372ee48f10c5"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet_mask.nii.gz:md5,8ab73e89ef64b280ff3ae93b84899bd9"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_boundingBox.pkl:md5,9433c088cb929add71ecbe76844628e9"
                    ]
                ],
                "3": [
                    "versions.yml:md5,5336fac953810299feeb1a62e824ad35"
                ],
                "bbox": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_boundingBox.pkl:md5,9433c088cb929add71ecbe76844628e9"
                    ]
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet.nii.gz:md5,687ccd83b5f7d49608e3372ee48f10c5"
                    ]
                ],
                "mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet_mask.nii.gz:md5,8ab73e89ef64b280ff3ae93b84899bd9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5336fac953810299feeb1a62e824ad35"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-23T16:36:43.652804"
    },
    "betcrop - fslbetcrop - t1 - dilate": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet.nii.gz:md5,687ccd83b5f7d49608e3372ee48f10c5"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet_mask.nii.gz:md5,8af0f9025e28e2c96a62e3a4cb01ab48"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_boundingBox.pkl:md5,9433c088cb929add71ecbe76844628e9"
                    ]
                ],
                "3": [
                    "versions.yml:md5,5336fac953810299feeb1a62e824ad35"
                ],
                "bbox": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_boundingBox.pkl:md5,9433c088cb929add71ecbe76844628e9"
                    ]
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet.nii.gz:md5,687ccd83b5f7d49608e3372ee48f10c5"
                    ]
                ],
                "mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_bet_mask.nii.gz:md5,8af0f9025e28e2c96a62e3a4cb01ab48"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5336fac953810299feeb1a62e824ad35"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-23T16:37:07.729622"
    }
}